core/shared/src/main/scala/zio/stm/TRandom.scala

```diff
@@ -0,0 +1,368 @@
+            def withSeed[A](f: Seed => (A, Seed)): STM[Nothing, A] =
+              seed.modify(f)
+
+            new Service {
+              def nextBoolean: STM[Nothing, Boolean] =
+                withSeed(rndBoolean)
+
+              def nextBytes(length: Int): STM[Nothing, Chunk[Byte]] =
+                withSeed(rndBytes(_, length))
+
+              def nextDouble: STM[Nothing, Double] =
+                withSeed(rndDouble)
+
+              def nextDoubleBetween(minInclusive: Double, maxExclusive: Double): STM[Nothing, Double] =
+                nextDoubleBetweenWith(minInclusive, maxExclusive)(nextDouble)
+
+              def nextFloat: STM[Nothing, Float] =
+                withSeed(rndFloat)
+
+              def nextFloatBetween(minInclusive: Float, maxExclusive: Float): STM[Nothing, Float] =
+                nextFloatBetweenWith(minInclusive, maxExclusive)(nextFloat)
+
+              def nextGaussian: STM[Nothing, Double] =
+                withSeed(rndGaussian)
+
+              def nextInt: STM[Nothing, Int] =
+                withSeed(rndInt)
+
+              def nextIntBetween(minInclusive: Int, maxExclusive: Int): STM[Nothing, Int] =
+                nextIntBetweenWith(minInclusive, maxExclusive)(nextInt, nextIntBounded)
+
+              def nextIntBounded(n: Int): STM[Nothing, Int] =
+                withSeed(rndInt(_, n))
+
+              def nextLong: STM[Nothing, Long] =
+                withSeed(rndLong)
+
+              def nextLongBetween(minInclusive: Long, maxExclusive: Long): STM[Nothing, Long] =
+                nextLongBetweenWith(minInclusive, maxExclusive)(nextLong, nextLongBounded)
+
+              def nextLongBounded(n: Long): STM[Nothing, Long] =
+                nextLongBoundedWith(n)(nextLong)
+
+              def nextPrintableChar: STM[Nothing, Char] =
+                withSeed(rndPrintableChar)
+
+              def nextString(length: Int): STM[Nothing, String] =
+                withSeed(rndString(_, length))
+
+              def setSeed(newSeed: Long): STM[Nothing, Unit] =
+                seed.set(newSeed)
+
+              def shuffle[A, Collection[+Element] <: Iterable[Element]](collection: Collection[A])(implicit
+                bf: BuildFrom[Collection[A], A, Collection[A]]
+              ): STM[Nothing, Collection[A]] =
+                shuffleWith(nextIntBounded(_), collection)
+            }
+          }
+          .commit
+      }
+    }
+
+  private[zio] def nextDoubleBetweenWith(minInclusive: Double, maxExclusive: Double)(
+    nextDouble: STM[Nothing, Double]
+  ): STM[Nothing, Double] =
+    if (minInclusive >= maxExclusive)
+      STM.die(new IllegalArgumentException("invalid bounds"))
+    else
+      nextDouble.map { n =>
+        val result = n * (maxExclusive - minInclusive) + minInclusive
+        if (result < maxExclusive) result
+        else Math.nextAfter(maxExclusive, Float.NegativeInfinity)
+      }
+
+  private[zio] def nextFloatBetweenWith(minInclusive: Float, maxExclusive: Float)(
```

Comment:
This is currently duplicated with `zio.random.Random`, but I don't think there is a good way to share this code sadly.

Reply:
Aren't they just in a static object so we can reference them?

Reply:
Ah but they return ZIO effects.